command/agent/monitor/stream_helpers.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package monitor
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/hashicorp/go-msgpack/v2/codec"
+	sframer "github.com/hashicorp/nomad/client/lib/streamframer"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// Stream Helpers
```

Comment:
Let's put some docstrings on this `StreamReader` and `StreamEncoder` just to give signposts for the next person who has to touch this code :grin: 

Ex. What's the difference between these two types? What piece of data is the mutex protecting and who is using that data concurrently?